File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/expression/MemberExpressionParsingFunction.java
package me.itzisonn_.meazy.parser.pasing_function.expression;

import me.itzisonn_.meazy.MeazyMain;
import me.itzisonn_.meazy.lexer.TokenTypeSets;
import me.itzisonn_.meazy.lexer.TokenTypes;
import me.itzisonn_.meazy.parser.ParsingContext;
import me.itzisonn_.meazy.lang.text.Text;
import me.itzisonn_.meazy.parser.Parser;
import me.itzisonn_.meazy.parser.UnexpectedTokenException;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.parser.ast.expression.Identifier;
import me.itzisonn_.meazy.parser.ast.expression.MemberExpression;
import me.itzisonn_.meazy.parser.ast.expression.CallExpression;
import me.itzisonn_.meazy.parser.pasing_function.AbstractParsingFunction;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public class MemberExpressionParsingFunction extends AbstractParsingFunction<Expression> {
    public MemberExpressionParsingFunction() {
        super("member_expression");
    }

    @Override
    public Expression parse(ParsingContext context, @Nullable Object... extra) {
        Parser parser = context.getParser();
        Expression object = parser.parseAfter(MeazyMain.getDefaultIdentifier("member_expression"), Expression.class);

        while (TokenTypeSets.MEMBER_ACCESS().contains(parser.getCurrent().getType())) {
            boolean isNullSafe = parser.getCurrentAndNext().getType().equals(TokenTypes.QUESTION_DOT());
            Expression member = parser.parseAfter(MeazyMain.getDefaultIdentifier("member_expression"), Expression.class);

            if (!(member instanceof Identifier) && !(member instanceof CallExpression)) {
                throw new UnexpectedTokenException(parser.getCurrent().getLine(), Text.translatable("meazy:parser.exception.member_expression"));
            }

            object = new MemberExpression(object, member, isNullSafe);
        }

        return object;
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/parser/pasing_function/expression/MemberExpressionParsingFunction.kt
package me.itzisonn_.meazy.parser.pasing_function.expression

import me.itzisonn_.meazy.lexer.TokenTypeSets.memberAccess
import me.itzisonn_.meazy.lexer.TokenTypes.questionDot
import me.itzisonn_.meazy.parser.Parser
import me.itzisonn_.meazy.parser.UnexpectedTokenException
import me.itzisonn_.meazy.parser.ast.expression.CallExpression
import me.itzisonn_.meazy.parser.ast.expression.Expression
import me.itzisonn_.meazy.parser.ast.expression.MemberExpression
import me.itzisonn_.meazy.parser.ast.expression.identifier.Identifier
import me.itzisonn_.meazy.parser.pasing_function.ParsingFunction
import me.itzisonn_.meazy.util.text.translatable

object MemberExpressionParsingFunction : ParsingFunction<Expression> {
    override fun Parser.parse(vararg extra: Any?): Expression {
        var receiver = parse(CallExpressionParsingFunction)

        while (isNext(memberAccess)) {
            val isNullSafe = consume(memberAccess, null).type == questionDot
            val member = parse(CallExpressionParsingFunction)

            if (member !is Identifier && member !is CallExpression) {
                throw UnexpectedTokenException(
                    current,
                    translatable("meazy:parser.exception.member_expression")
                )
            }

            receiver = MemberExpression(receiver, member, isNullSafe)
        }

        return receiver
    }
}
